{
 "metadata": {
  "name": "",
  "signature": "sha256:2ca6a54ff347c0c99c9120806d55e4e87c5d425b89f23a031ab2e5c6602b505a"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%matplotlib inline\n",
      "import csv\n",
      "import gzip\n",
      "import numpy as np\n",
      "import itertools\n",
      "import pandas\n",
      "import json\n",
      "from sklearn import linear_model, model_selection, metrics\n",
      "\n",
      "import matplotlib.pyplot as plt\n",
      "import matplotlib.gridspec as gridspec\n",
      "from matplotlib.backends.backend_pdf import PdfPages\n",
      "\n",
      "from adjustText import adjust_text"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "validation_file = \"Data/mc3_all_pairsets.aggregated.maf\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\"\"\"\n",
      "germline_data = []\n",
      "validated_data = []\n",
      "unvalidated_data = []\n",
      "with open(validation_file) as handle:\n",
      "    reader = csv.DictReader(handle, delimiter=\"\\t\")\n",
      "    for row in reader:\n",
      "        s = row['mutval_status_targeted']\n",
      "        if s == \"germline_powered\":\n",
      "            germline_data.append(row)\n",
      "        if s == \"validated_powered\":\n",
      "            validated_data.append(row)\n",
      "        if s == \"unvalidated_powered\":\n",
      "            unvalidated_data.append(row)\n",
      "\"\"\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 3,
       "text": [
        "'\\ngermline_data = []\\nvalidated_data = []\\nunvalidated_data = []\\nwith open(validation_file) as handle:\\n    reader = csv.DictReader(handle, delimiter=\"\\t\")\\n    for row in reader:\\n        s = row[\\'mutval_status_targeted\\']\\n        if s == \"germline_powered\":\\n            germline_data.append(row)\\n        if s == \"validated_powered\":\\n            validated_data.append(row)\\n        if s == \"unvalidated_powered\":\\n            unvalidated_data.append(row)\\n'"
       ]
      }
     ],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "germline_data = []\n",
      "validated_data = []\n",
      "unvalidated_data = []\n",
      "\n",
      "val_order = [\n",
      "    \"mutval_status_targeted\",\n",
      "    \"mutval_status_wgs\",\n",
      "    #\"mutval_status_rna\",\n",
      "    \"mutval_status_lowpass\"\n",
      "]\n",
      "\n",
      "def row_extract(row):\n",
      "    out = {}\n",
      "    for i in [\"FILTER\", 'Variant_Type', 'CENTERS']:\n",
      "        out[i] = row[i]\n",
      "    return out\n",
      "\n",
      "data_types = ['targeted','wgs']\n",
      "\n",
      "def validation_status(line):\n",
      "    consensus_judgement = 0\n",
      "    powered = False\n",
      "    for d in data_types:\n",
      "        power_field = 'validation_power_' + d\n",
      "        d_power = line[power_field]\n",
      "        if d_power != \"\" and float(d_power) < 0.95:\n",
      "            continue\n",
      "        judgement_field = 'validation_judgement_' + d\n",
      "        if line[judgement_field] == '' or line[judgement_field] == 'NaN':\n",
      "            continue\n",
      "        powered = True\n",
      "\n",
      "        d_judgement = int(line[judgement_field])\n",
      "        # 0 = unvalidated, 1 = validated, 2 = germline\n",
      "        if d_judgement > consensus_judgement:\n",
      "            consensus_judgement = d_judgement\n",
      "    if not powered:\n",
      "        return None\n",
      "    if consensus_judgement == 0:\n",
      "        return \"unvalidated_powered\"\n",
      "    if consensus_judgement == 1:\n",
      "        return \"validated_powered\"\n",
      "    if consensus_judgement == 2:\n",
      "        return \"germline_powered\"\n",
      "\n",
      "with open(validation_file) as handle:\n",
      "    reader = csv.DictReader(handle, delimiter=\"\\t\")\n",
      "    for row in reader:\n",
      "        if row['FILTER'].count(\"bitgt\") == 0:\n",
      "            s = validation_status(row)\n",
      "            if s == \"germline_powered\":\n",
      "                germline_data.append(row_extract(row))\n",
      "            if s == \"validated_powered\":\n",
      "                validated_data.append(row_extract(row))\n",
      "            if s == \"unvalidated_powered\":\n",
      "                unvalidated_data.append(row_extract(row))\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print len(germline_data)\n",
      "print len(validated_data)\n",
      "print len(unvalidated_data)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "805\n",
        "468377\n",
        "65846\n"
       ]
      }
     ],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "snp_callers = [\"SOMATICSNIPER\", \"RADIA\", \"MUTECT\", \"MUSE\", \"VARSCANS\"]\n",
      "indel_callers = [\"PINDEL\", \"INDELOCATOR\", \"VARSCANS\"]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def maf_row_valid(row, exons_only=False, snp=True):\n",
      "    if snp:\n",
      "        if row['Variant_Type'] != 'SNP':\n",
      "            return False\n",
      "    else:\n",
      "        if row['Variant_Type'] == 'SNP':\n",
      "            return False\n",
      "    f = row['FILTER'].split(\",\")\n",
      "    if exons_only:\n",
      "        if \"NonExonic\" in f or \"bitgt\" in f:\n",
      "            return False\n",
      "    for i in ['gapfiller', 'wga', 'contest', 'native_wga_mix', 'NonExonic', 'nonpreferredpair']:\n",
      "        if i in f:\n",
      "            return False\n",
      "    return True\n",
      "\n",
      "def maf_row_cutoff(row, n=2):\n",
      "    return len(row[\"CENTERS\"].split(\"|\")) >= n\n",
      "    \"\"\"\n",
      "    count = 0\n",
      "    for c in row[\"CENTERS\"].split(\"|\"):\n",
      "        c = c.replace(\"*\", \"\")\n",
      "        if c in [\"SOMATICSNIPER\", \"RADIA\", \"MUTECT\", \"MUSE\", \"VARSCANS\"]:\n",
      "            count += 1\n",
      "    return count >= n\n",
      "    \"\"\"\n",
      "\n",
      "def maf_row_caller(row, caller):\n",
      "    s = row[\"CENTERS\"].replace(\"*\", \"\", -1).split(\"|\")\n",
      "    return caller in s\n",
      "\n",
      "def maf_row_filter(row, filter):\n",
      "    if filter == \"NONE\":\n",
      "        return True\n",
      "    s = row['FILTER'].split(',')\n",
      "    if filter == \"PASS\":\n",
      "        if row['FILTER'] == \"PASS\":\n",
      "            return True\n",
      "        if \"NonExonic\" in s:\n",
      "            f.remove(\"NonExonic\")\n",
      "        if \"bitgt\" in s:\n",
      "            f.remove(\"bitgt\")\n",
      "        if len(s) == 0:\n",
      "            return True\n",
      "        return False\n",
      "    return filter not in s"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_stats(germline_data, unvalidated_data, validated_data, valid, compare):\n",
      "    cfm = {\n",
      "        \"fp\" : 0.0,\n",
      "        \"fn\" : 0.0,\n",
      "        \"tp\" : 0.0,\n",
      "        \"tn\" : 0.0\n",
      "    }\n",
      "    germline_count = 0\n",
      "    validated_count = 0\n",
      "    unvalidated_count = 0\n",
      "    missed_count = 0\n",
      "    for row in germline_data:\n",
      "        if valid(row):\n",
      "            if compare(row):\n",
      "                cfm[\"fp\"] += 1\n",
      "                germline_count += 1\n",
      "            else:\n",
      "                cfm[\"tn\"] += 1\n",
      "    for row in unvalidated_data:\n",
      "        if valid(row):\n",
      "            if compare(row):\n",
      "                cfm[\"fp\"] += 1\n",
      "                unvalidated_count += 1\n",
      "            else:\n",
      "                cfm[\"tn\"] += 1\n",
      "    for row in validated_data:\n",
      "        if valid(row):\n",
      "            if compare(row):\n",
      "                cfm[\"tp\"] += 1\n",
      "                validated_count += 1\n",
      "            else:\n",
      "                cfm[\"fn\"] += 1\n",
      "                missed_count += 1\n",
      "    return {\n",
      "        \"germline_count\" : germline_count,\n",
      "        \"validated_count\" : validated_count,\n",
      "        \"unvalidated_count\" : unvalidated_count,\n",
      "        \"missed_count\" : missed_count,\n",
      "        \"matrix\" : cfm\n",
      "    }\n",
      "   "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "isect_snp_data = []\n",
      "isect_indel_data = []\n",
      "    \n",
      "for i in range(1, 6):\n",
      "    snp_data = get_stats(germline_data, unvalidated_data, validated_data, \n",
      "              lambda x: maf_row_valid(x, True, True), lambda x: maf_row_cutoff(x,i) and maf_row_filter(x,\"PASS\"))\n",
      "    isect_snp_data.append( (\"%d\" % i, snp_data) )\n",
      "\n",
      "for i in range(1, 4):\n",
      "    indel_data = get_stats(germline_data, unvalidated_data, validated_data, \n",
      "          lambda x: maf_row_valid(x, True, False), lambda x: maf_row_cutoff(x,i) and maf_row_filter(x,\"PASS\"))\n",
      "    isect_indel_data.append( (\"%d\" % i, indel_data) )\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "caller_snp_data = []\n",
      "\n",
      "for c in snp_callers:\n",
      "    data = get_stats(germline_data, unvalidated_data, validated_data, \n",
      "              lambda x: maf_row_valid(x, True, True), lambda x: maf_row_caller(x,c) and maf_row_filter(x,\"PASS\"))\n",
      "    caller_snp_data.append( (c, data) )\n",
      "\n",
      "caller_indel_data = []\n",
      "\n",
      "for c in indel_callers:\n",
      "    data = get_stats(germline_data, unvalidated_data, validated_data, \n",
      "              lambda x: maf_row_valid(x, True, False), lambda x: maf_row_caller(x,c) and maf_row_filter(x,\"PASS\"))\n",
      "    caller_indel_data.append( (c, data) )\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 35
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "caller_isect = []\n",
      "for i, a in enumerate(snp_callers):\n",
      "    for j, b in enumerate(snp_callers):\n",
      "        if i > j:\n",
      "            data = get_stats(germline_data, unvalidated_data, validated_data, \n",
      "              lambda x: maf_row_valid(x, True, True), lambda x: maf_row_caller(x,a) and maf_row_caller(x,b))\n",
      "            caller_isect.append( (\"%s and %s\" % (a,b), data) )\n",
      "\n",
      "caller_isect.sort(key=lambda x: x[1]['validated_count'], reverse=True)        "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 11
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "filter_order = [\n",
      "    'NONE',\n",
      "    'PASS',\n",
      "    'broad_PoN_v2',\n",
      "    #'NonExonic',\n",
      "    'StrandBias',\n",
      "    #'badseq',\n",
      "    #'bitgt',\n",
      "    'common_in_exac',\n",
      "    #'contest',\n",
      "    'ndp',\n",
      "    #'nonpreferredpair',\n",
      "    'oxog',\n",
      "    #'wga'\n",
      "]\n",
      "\n",
      "filter_data = []\n",
      "for f in filter_order:\n",
      "    data = get_stats(germline_data, unvalidated_data, validated_data, \n",
      "              lambda x: maf_row_valid(x, True, True), lambda x: maf_row_filter(x,f) )\n",
      "    filter_data.append( (f, data) )\n",
      "filter_data.sort(key=lambda x: x[1]['validated_count'], reverse=True)        "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 12
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def make_bars(datasets):\n",
      "    validated = []\n",
      "    germline = []\n",
      "    unvalidated = []\n",
      "    false_negative = []\n",
      "    true_negative = []\n",
      "    \n",
      "    for d in datasets:\n",
      "        validated.append( d['validated_count'] )\n",
      "        unvalidated.append( d['unvalidated_count'] )\n",
      "        germline.append( d['germline_count'] )\n",
      "        false_negative.append( d['matrix']['fn'] )\n",
      "        true_negative.append( d['matrix']['tn'] )\n",
      "    return {\n",
      "        'validated' : np.array(validated),\n",
      "        'unvalidated' : np.array(unvalidated),\n",
      "        'germline' : np.array(germline),\n",
      "        'false_negative' : np.array(false_negative),\n",
      "        'true_negative' : np.array(true_negative)\n",
      "    }"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 13
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "TP_COLOR = \"#5DA5A1\"\n",
      "GERM_COLOR = \"#C45331\"\n",
      "UNVAL_COLOR = \"#E7960A\"\n",
      "FN_COLOR = \"#F6E849\"\n",
      "TN_COLOR = \"#016384\"\n",
      "\n",
      "pdf = PdfPages('Pdfs/Figure_4.pdf')\n",
      "\n",
      "width = 0.75\n",
      "fig = plt.figure(figsize=(15, 25))\n",
      "plt.title('Error Profile vs Caller Overlap')\n",
      "\n",
      "caller_bars = make_bars(list(a[1] for a in caller_snp_data))\n",
      "ind = np.arange(len(caller_snp_data))\n",
      "ax2 = plt.subplot2grid((4,2),(0,0))\n",
      "\n",
      "ax2.bar(ind, caller_bars['validated'], width, color=TP_COLOR)\n",
      "ax2.bar(ind, caller_bars['germline'], width, bottom=caller_bars['validated'], color=GERM_COLOR)\n",
      "ax2.bar(ind, caller_bars['unvalidated'], width, bottom=caller_bars['validated'] + caller_bars['germline'], color=UNVAL_COLOR)\n",
      "ax2.bar(ind, caller_bars['false_negative'], width, \n",
      "             bottom=caller_bars['validated'] + caller_bars['germline'] + caller_bars['unvalidated'], \n",
      "             color=FN_COLOR)\n",
      "#ax2.bar(ind, caller_bars['true_negative'], width, \n",
      "#             bottom=caller_bars['validated'] + caller_bars['germline'] + caller_bars['unvalidated'] + caller_bars['false_negative'], \n",
      "#             color=TN_COLOR)\n",
      "ax2.set_xticks(range(len(caller_bars)))\n",
      "\n",
      "ax2.set_xticks(range(len(caller_snp_data)))\n",
      "ax2.set_xticklabels(list(i[0] for i in caller_snp_data), rotation=15)\n",
      "\n",
      "ax1 = plt.subplot2grid((4,2),(1,0))\n",
      "ax1.set_title(\"SNP Caller Overlap\")\n",
      "\n",
      "snp_bars = make_bars(list(a[1] for a in isect_snp_data))\n",
      "ind = np.arange(len(isect_snp_data))\n",
      "\n",
      "p1 = ax1.bar(ind, snp_bars['validated'], width, color=TP_COLOR)\n",
      "p2 = ax1.bar(ind, snp_bars['germline'], width, bottom=snp_bars['validated'], color=GERM_COLOR)\n",
      "p3 = ax1.bar(ind, snp_bars['unvalidated'], width, bottom=snp_bars['validated'] + snp_bars['germline'], color=UNVAL_COLOR)\n",
      "p4 = ax1.bar(ind, snp_bars['false_negative'], width, \n",
      "             bottom=snp_bars['validated'] + snp_bars['germline'] + snp_bars['unvalidated'], \n",
      "             color=FN_COLOR)\n",
      "#p5 = ax1.bar(ind, snp_bars['true_negative'], width, \n",
      "#             bottom=snp_bars['validated'] + snp_bars['germline'] + snp_bars['unvalidated'] + snp_bars['false_negative'], \n",
      "#             color=TN_COLOR)\n",
      "\n",
      "ax1.set_xticks(range(len(isect_snp_data)))\n",
      "ax1.set_xticklabels(list(i[0] for i in isect_snp_data))\n",
      "\n",
      "\n",
      "ax2 = plt.subplot2grid((4,2),(1,1))\n",
      "ax2.set_title(\"INDEL Caller Overlap\")\n",
      "ind = np.arange(len(isect_indel_data))\n",
      "\n",
      "indel_bars = make_bars(list(a[1] for a in isect_indel_data))\n",
      "\n",
      "ax2.bar(ind, indel_bars['validated'], width, color=TP_COLOR)\n",
      "ax2.bar(ind, indel_bars['germline'], width, bottom=indel_bars['validated'], color=GERM_COLOR)\n",
      "ax2.bar(ind, indel_bars['unvalidated'], width, bottom=indel_bars['validated'] + indel_bars['germline'], color=UNVAL_COLOR)\n",
      "ax2.bar(ind, indel_bars['false_negative'], width, \n",
      "             bottom=indel_bars['validated'] + indel_bars['germline'] + indel_bars['unvalidated'], \n",
      "             color=FN_COLOR)\n",
      "#ax2.bar(ind, indel_bars['true_negative'], width, \n",
      "#             bottom=indel_bars['validated'] + indel_bars['germline'] + indel_bars['unvalidated'] + indel_bars['false_negative'], \n",
      "#             color=TN_COLOR)\n",
      "ax2.set_xticks(range(len(isect_indel_data)))\n",
      "\n",
      "ax2.set_xticklabels(list(i[0] for i in isect_indel_data))\n",
      "\n",
      "#plt.xticks(ind , ('1', '2', '3', '4', '5'))\n",
      "#plt.xlabel(\"Caller Overlap\")\n",
      "\n",
      "ax2 = plt.subplot2grid((4,2),(0,1))\n",
      "ax2.axes.get_xaxis().set_visible(False)\n",
      "ax2.axes.get_yaxis().set_visible(False)\n",
      "ax2.axes.get_xaxis().set_visible(False)\n",
      "ax2.axis('off')\n",
      "ax2.patch.set_visible(False)\n",
      "leg = ax2.legend((p1[0], p2[0], p3[0], p4[0]), #, p5[0]), \n",
      "           ('True Positive', 'False Positive(Germline)', \"False Positive (UnValidated)\", \n",
      "            \"FalseNegative\"), #, \"TrueNegative\"),\n",
      "               prop={'size':20})\n",
      "leg.set_title('Call Validation Status',prop={'size':20} )\n",
      "\n",
      "\n",
      "ax = plt.subplot2grid((4,2),(2,0), colspan=3)\n",
      "\n",
      "caller_bars = make_bars(list(a[1] for a in caller_isect))\n",
      "width = 0.75\n",
      "ind = np.arange(len(caller_isect))\n",
      "\n",
      "ax.bar(ind, caller_bars['validated'], width, color=TP_COLOR)\n",
      "ax.bar(ind, caller_bars['germline'], width, bottom=caller_bars['validated'], color=GERM_COLOR)\n",
      "ax.bar(ind, caller_bars['unvalidated'], width, bottom=caller_bars['validated'] + caller_bars['germline'], color=UNVAL_COLOR)\n",
      "ax.bar(ind, caller_bars['false_negative'], width, \n",
      "             bottom=caller_bars['validated'] + caller_bars['germline'] + caller_bars['unvalidated'], \n",
      "             color=FN_COLOR)\n",
      "#ax.bar(ind, caller_bars['true_negative'], width, \n",
      "#             bottom=caller_bars['validated'] + caller_bars['germline'] + caller_bars['unvalidated'] + caller_bars['false_negative'], \n",
      "#             color=TN_COLOR)\n",
      "\n",
      "ax.set_xticks(range(len(caller_isect)))\n",
      "ax.set_xticklabels(list(i[0] for i in caller_isect), rotation=15)\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "filter_bars = make_bars(list(a[1] for a in filter_data))\n",
      "width = 0.75\n",
      "ind = np.arange(len(filter_data))\n",
      "\n",
      "ax = plt.subplot2grid((4,2),(3,0), colspan=3)\n",
      "\n",
      "ax.bar(ind, filter_bars['validated'], width, color=TP_COLOR)\n",
      "ax.bar(ind, filter_bars['germline'], width, bottom=filter_bars['validated'], color=GERM_COLOR)\n",
      "ax.bar(ind, filter_bars['unvalidated'], width, bottom=filter_bars['validated'] + filter_bars['germline'], color=UNVAL_COLOR)\n",
      "ax.bar(ind, filter_bars['false_negative'], width, \n",
      "             bottom=filter_bars['validated'] + filter_bars['germline'] + filter_bars['unvalidated'], \n",
      "             color=FN_COLOR)\n",
      "#ax.bar(ind, filter_bars['true_negative'], width, \n",
      "#             bottom=filter_bars['validated'] + filter_bars['germline'] + filter_bars['unvalidated'] + filter_bars['false_negative'], \n",
      "#             color=TN_COLOR)\n",
      "\n",
      "ax.set_xticks(range(len(filter_data)))\n",
      "ax.set_xticklabels(list(i[0] for i in filter_data), rotation=15)\n",
      "\n",
      "fig.text(0.05, .9, \"A)\", fontsize=20)\n",
      "fig.text(0.05, .70, \"B)\", fontsize=20)\n",
      "fig.text(0.05, .50, \"C)\", fontsize=20)\n",
      "fig.text(0.05, .30, \"D)\", fontsize=20)\n",
      "\n",
      "txt = \"\"\"Figure 4: Validation Statistics. A) The composition of validation status for call from each of the independent SNP callers.\n",
      "B) Composition of variants with overlapping callers after application of standard filters. Starting with any caller and increasing\n",
      "to require more callers agree on a site. C) Composition of validation status for pairs of callers. D) Composition of validation \n",
      "status for each of the filters, and finally 'PASS' which represents sites which pass all filters.\n",
      "\"\"\"\n",
      "fig.text(.05,.05, txt, fontdict={'size' : 14})\n",
      "\n",
      "#plt.tight_layout(pad=0.4)\n",
      "\n",
      "pdf.savefig()\n",
      "plt.close()\n",
      "pdf.close()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 14
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from IPython.display import FileLink, FileLinks\n",
      "FileLink('Pdfs/Figure_4.pdf')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<a href='Pdfs/Figure_4.pdf' target='_blank'>Pdfs/Figure_4.pdf</a><br>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 15,
       "text": [
        "/mnt/mc3-analysis/Pdfs/Figure_4.pdf"
       ]
      }
     ],
     "prompt_number": 15
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(caller_snp_data)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 16,
       "text": [
        "5"
       ]
      }
     ],
     "prompt_number": 16
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "out = []\n",
      "for i, v in caller_snp_data:\n",
      "    o = dict(name=i, \n",
      "             unvalidated_count=v['unvalidated_count'], \n",
      "             validated_count=v['validated_count'],\n",
      "             germline_count=v['germline_count'],\n",
      "             false_positive = v['matrix']['fp'],\n",
      "             false_negative = v['matrix']['fn'],\n",
      "             true_positive  = v['matrix']['tp'],\n",
      "             true_negative  = v['matrix']['tn'],\n",
      "             )\n",
      "    out.append(o)\n",
      "\n",
      "for i, v in caller_indel_data:\n",
      "    o = dict(name=i, \n",
      "             unvalidated_count=v['unvalidated_count'], \n",
      "             validated_count=v['validated_count'],\n",
      "             germline_count=v['germline_count'],\n",
      "             false_positive = v['matrix']['fp'],\n",
      "             false_negative = v['matrix']['fn'],\n",
      "             true_positive  = v['matrix']['tp'],\n",
      "             true_negative  = v['matrix']['tn'],\n",
      "             )\n",
      "    out.append(o)\n",
      "\n",
      "\n",
      "for i, v in isect_snp_data:\n",
      "    o = dict(name=\"%s SNP Callers\" % i, \n",
      "             unvalidated_count=v['unvalidated_count'], \n",
      "             validated_count=v['validated_count'],\n",
      "             germline_count=v['germline_count'],\n",
      "             false_positive = v['matrix']['fp'],\n",
      "             false_negative = v['matrix']['fn'],\n",
      "             true_positive  = v['matrix']['tp'],\n",
      "             true_negative  = v['matrix']['tn'],\n",
      "             )\n",
      "    out.append(o)\n",
      "\n",
      "for i, v in isect_indel_data:\n",
      "    o = dict(name=\"%s INDEL Callers\" % i, \n",
      "             unvalidated_count=v['unvalidated_count'], \n",
      "             validated_count=v['validated_count'],\n",
      "             germline_count=v['germline_count'],\n",
      "             false_positive = v['matrix']['fp'],\n",
      "             false_negative = v['matrix']['fn'],\n",
      "             true_positive  = v['matrix']['tp'],\n",
      "             true_negative  = v['matrix']['tn'],\n",
      "             )\n",
      "    out.append(o)\n",
      "\n",
      "for i, v in filter_data:\n",
      "    o = dict(name=\"Filter=%s\" % i, \n",
      "             unvalidated_count=v['unvalidated_count'], \n",
      "             validated_count=v['validated_count'],\n",
      "             germline_count=v['germline_count'],\n",
      "             false_positive = v['matrix']['fp'],\n",
      "             false_negative = v['matrix']['fn'],\n",
      "             true_positive  = v['matrix']['tp'],\n",
      "             true_negative  = v['matrix']['tn'],\n",
      "             )\n",
      "    out.append(o)\n",
      "\n",
      "    \n",
      "for i, v in caller_isect:\n",
      "    o = dict(name=i, \n",
      "             unvalidated_count=v['unvalidated_count'], \n",
      "             validated_count=v['validated_count'],\n",
      "             germline_count=v['germline_count'],\n",
      "             false_positive = v['matrix']['fp'],\n",
      "             false_negative = v['matrix']['fn'],\n",
      "             true_positive  = v['matrix']['tp'],\n",
      "             true_negative  = v['matrix']['tn'],\n",
      "             )\n",
      "    out.append(o)\n",
      "\n",
      "table_data = pandas.DataFrame(out)[['name', 'validated_count', \n",
      "                             'unvalidated_count', 'germline_count', 'true_positive', 'true_negative',\n",
      "                             'false_positive', 'false_negative']]\n",
      "#print caller_snp_data\n",
      "#print caller_isect\n",
      "#print filter_data\n",
      "#print isect_indel_data\n",
      "#print isect_snp_data"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 44
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "table_data.to_csv(\"Figure4.tsv\", sep=\"\\t\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 45
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "FileLink('Figure4.tsv')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<a href='Figure4.tsv' target='_blank'>Figure4.tsv</a><br>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 46,
       "text": [
        "/mnt/mc3-analysis/Figure4.tsv"
       ]
      }
     ],
     "prompt_number": 46
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}